### ISSUE TITLE
SC2178/SC2128 false positives on local variables

### ISSUE THREAD

**Issue by knirch** [koalaman/shellcheck#680]
```bash
#!/bin/bash

foo () {
   local baz=()
   echo "${baz[@]}"
}

bar () {
   local baz=""
   # Variable was used as an array but is now assigned a string. [SC2178]
   echo "$baz"
   # Expanding an array without an index only gives the first element. [SC2128]
}
```

**Comment by knirch** [comment]
Saw the mention on the [SC2178](https://github.com/koalaman/shellcheck/wiki/Sc2178) page. Close if you want to, not too much refactoring needed to work around this.

**Comment by knirch** [comment]
Unsure if things have changed, but whilst clearing up old warnings I noticed that the above code can be fixed by simply changing local baz="" to local baz

```sh
#!/bin/bash

foo () {                                                                                                                                                       
   local baz=()
   echo "${baz[@]}"
}

bar () {
   local baz
   echo "$baz"
}
```

**Comment by b0o** [comment]
@knirch Thanks for describing that workaround, I updated the wiki to include it: [SC2128](https://github.com/koalaman/shellcheck/wiki/SC2128), [SC2178](https://github.com/koalaman/shellcheck/wiki/SC2178)

**Comment by ingydotnet** [comment]
Also functions defined using outer parens rather than braces like:
```
do-it() (
  foo=(xyz)
)

do-also() (
  foo=xyz
)
```
should not trigger these warnings.

I use parens for defining functions as much as possible as it limits global side-effects. 
Vars defined within are essentially "local".
